src/main/java/com/ibm/darpc/DaRPCMemPool.java

```diff
@@ -0,0 +1,435 @@
+		}
+		if (hugePageLimit >= 0) {
+			this.hugePageLimit = hugePageLimit;
+		}
+		if ((hugePagePath != null) && (!hugePagePath.equals("")) && (this.hugePageLimit > 0)) {
+			dir = new File(hugePagePath);
+			if (!dir.exists()){
+				dir.mkdirs();
+			}
+			for (File child : dir.listFiles()) {
+				child.delete();
+			}
+		} else {
+			this.hugePageLimit = 0;
+		}
+		pdMemPool = new HashMap<IbvPd, PdMemPool>();
+		mrs = new LinkedList<IbvMr>();
+		isOpen = true;
+	}
+
+	public synchronized void close() {
+		if (isOpen) {
+			isOpen = false;
+			pdMemPool = null;
+			for (Iterator<IbvMr> it = mrs.iterator(); it.hasNext(); ) {
+				IbvMr m = it.next();
+				try {
+					m.deregMr().execute().free();
+				} catch (IOException e) {
+					System.out.println("Could not unregister memory region.");
+					e.printStackTrace();
+				}
+			}
+			mrs = null;
+		}
+
+		// If hugepages were used, clean and delete the created files and the directory.
+		if (hugePageLimit > 0) {
+			if (dir.exists()) {
+				for (File child : dir.listFiles()) {
+					child.delete();
+				}
+				dir.delete();
+			}
+		}
+	}
+
+	public void finalize() {
+		// Just in case the user did not do that.
+		close();
+	}
+
+
+	// the next two methods allocate a buffer from the OS. The first one
+	// allocates from the regular heap and the second one from huge pages.
+	// These are two alternatives. If hugepages is configured by the user,
+	// memory will first be allocated from huge pages and after reacing the limit,
+	// more memory will be allocated from the regular heap.
+
+	// allocate a buffer from the regular heap
+	ByteBuffer allocateHeapBuffer() {
+		ByteBuffer byteBuffer;
+
+		if (alignmentSize > 1) {
+			ByteBuffer rawBuffer = ByteBuffer.allocateDirect(allocationSize + alignmentSize);
+			long rawBufferAddress = ((sun.nio.ch.DirectBuffer)rawBuffer).address();
+			long alignmentOffset = rawBufferAddress % alignmentSize;
+			if (alignmentOffset != 0) {
+				rawBuffer.position(alignmentSize - (int)alignmentOffset);
+			}
+			byteBuffer = rawBuffer.slice();
+
+		} else {
+			byteBuffer = ByteBuffer.allocateDirect(allocationSize);
+		}
+		return (byteBuffer);
+	}
+
+	// allocate a buffer from hugepages
+	ByteBuffer allocateHugePageBuffer() throws IOException {
+		String path = hugePagePath + "/" + currentRegion++ + ".mem";
+		RandomAccessFile randomFile = null;
+		try {
+			randomFile = new RandomAccessFile(path, "rw");
+		} catch (FileNotFoundException e) {
+			System.out.println("Path " + path + " to huge page directory not found.");
+			throw e;
+		}
+		try {
+			randomFile.setLength(allocationSize);
+		} catch (IOException e) {
+			System.out.println("Coult not set allocation length of mapped random access file on huge page directory.");
+			randomFile.close();
+			throw e;
+		}
+		FileChannel channel = randomFile.getChannel();
+		MappedByteBuffer mappedBuffer = null;
+		try {
+			mappedBuffer = channel.map(MapMode.READ_WRITE, 0,
+					allocationSize);
```

Comment:
what about alignment here? We only know that it is at least hugepage aligned but what if the user wants a bigger alignment requirement?